internal/flowrate/xrate/monitor.go

```diff
@@ -0,0 +1,98 @@
+package xrate
+
+import (
+	"time"
+
+	"golang.org/x/time/rate"
+
+	legacyimpl "github.com/cometbft/cometbft/internal/flowrate/legacy"
+	cmtsync "github.com/cometbft/cometbft/libs/sync"
+)
+
+type Monitor struct {
+	limiter *rate.Limiter
+
+	// rate.Limiter configuration
+	burst int
+	rate  int64
+
+	// Tokens reserved but not yet used
+	reserved int
+
+	statusMtx cmtsync.Mutex
+	status    legacyimpl.Status
+}
+
+func New(_, _ time.Duration) *Monitor {
+	return &Monitor{
+		limiter: rate.NewLimiter(rate.Inf, 0),
+	}
+}
+
+func (m *Monitor) Limit(want int, maxRate int64, block bool) (n int) {
+	// Update rate limiter parameters, if needed
+	// XXX: this should never change after the first call
+	if maxRate != m.rate {
+		m.limiter.SetLimit(rate.Limit(maxRate))
+		m.rate = maxRate
+	}
+	if want > m.burst {
+		m.limiter.SetBurst(want)
+		m.burst = want
+	}
+
+	var sleepTime time.Duration
+	for want > m.reserved && block {
+		tokens := want - m.reserved
+		// We cannot reserve more than m.limiter.Burst()
+		if tokens > m.burst {
+			tokens = m.burst
+		}
+
+		res := m.limiter.ReserveN(time.Now(), tokens)
+		if !res.OK() {
+			panic("ReserveN returned false")
+		}
+
+		// Requested tokens are available after res.Delay()
+		delay := res.Delay()
+		if delay > 0 {
+			start := time.Now()
+			time.Sleep(delay)
+			sleepTime += time.Since(start)
+		}
```

Comment:
Wouldn't it be simpler (and more precise) if we accumulated all the `delay` values that we're getting from the limiter and did just one sleep just after the loop?

Reply:
Yes, it would be simpler. But not, it would be less precise.

I've tried both.